src/main/java/frc/robot/subsystems/targetvision/TargetVision.kt

```diff
@@ -1,61 +1,203 @@
-    private val targetsByDistance: List<Pair<LimelightTarget_Retro, Double>>
-        get() {
-            return inputs.targets.map { target: LimelightTarget_Retro -> Pair(target, distanceToTarget(target)) }
-                .sortedBy { it.second }
+   data class Measurement(val timestamp: Double, val pose: LimelightTarget_Detector)
+
+    data class Sample(val timestamp : Double, val pose: Translation2d)
+
+    private fun takeMeasurement(target: LimelightTarget_Detector): Measurement{
+
+       //val targetTranslation = Translation2d(getDistance(target), target.tx)
+
+       return Measurement(inputs.lastTimeStampMS, target)
+
+   }
+
+
+
+    val primaryTargetSamples: List<Sample>
+        get(){
+            return robotPoses[0].map { pair -> Sample(pair.timestamp, Translation2d(getDistance(pair.pose), pair.pose.tx))}
+        }
+
+
+    val primaryTarget: LimelightTarget_Detector
+        get(){
+            return robotPoses[0].last().pose
         }
 
-    private var measurements: MutableList<Measurement> = mutableListOf()
 
-    data class Measurement(val timestamp: Double, val translation: Translation2d)
+    var robotPoses: MutableList<MutableList<Measurement>> = mutableListOf()
+
+
+    private fun groupConfidence(last: LimelightTarget_Detector, measurement: LimelightTarget_Detector): Double
+    {
+        var score = 0.0
 
-    private fun takeMeasurement(target: LimelightTarget_Retro): Measurement {
-        val targetTranslation = Translation2d(distanceToTarget(target), target.tx)
+        val deltaX = abs(measurement.tx - last.tx)
+        val deltaY = abs(measurement.ty - last.ty)
+        val deltaA = abs(measurement.ta.pow(0.5) - last.ta.pow(0.5))
 
-        return Measurement(timestamp = inputs.lastUpdateTimestamp, translation = targetTranslation)
+        score += (TX_WEIGHT)*( 1 - (deltaX.pow(TX_DROPOFF_RATE) / 54.0.pow(TX_DROPOFF_RATE)) ) //27 = horizontal fov of limelight
+        score += (TY_WEIGHT)*( 1 - (deltaY.pow(TY_DROPOFF_RATE) / 41.0.pow(TY_DROPOFF_RATE)) ) //20.5 = vertical fov of limelight
+        score += (TA_WEIGHT)*( 1 - (deltaA.pow(TA_DROPOFF_RATE) / TA_MAX.pow(TA_DROPOFF_RATE)) )
+
+        return score
     }
 
-    private fun addMeasurement(measurement: Measurement) {
-        measurements.add(measurement)
 
-        if (measurements.size > SAMPLE_NUM) {
-            measurements.removeFirst()
+    fun addMeasurement(measurement: Measurement){
+
+        var grouped = false
+
+        // group targets
+        for (pose in robotPoses) {
+            // TODO: Find good threshold values for grouping
+            if(groupConfidence(pose.last().pose, measurement.pose) > GROUPING_CONFIDENCE_THRESHOLD) {
+                pose.add(measurement)
+                if ((measurement.timestamp - pose.first().timestamp > MAX_SAMPLE_TIME_MS)) { // TODO: find good value
+                    // cleanup old targets
+                    pose.removeFirst()
+                }
+                grouped = true
```

Comment:
You could also just return and then make a new pose history unconditionally after the loop, removing the need for `grouped`.